### PR TITLE
Enable performance log checkbox menu item

### PR DESCRIPTION
**Pull request by DanVanAtta**
Adding "Perf Console", based on #333 (4 commits)

In this PR a checkbox is added to the game menu to 'enable performance logging'. This opens up a console window, the "performance console" that displays performance timing. The implementation borrows from Console window and looks similar.

[Review on Reviewable](https://reviewable.io/reviews/triplea-game/triplea/335)

### CODE REVIEW

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like this needs rebased.

**Comment by DanVanAtta** [comment]
Merge conflicts fixed.

**Comment by gaborbernat** [comment]
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/games/performance/Perf.java, line 41 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n3miiXwk5Nkj6WXsK:-K8n3miiXwk5Nkj6WXsL:-1304849619) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/performance/Perf.java#L41)):*
I think we should refactor this into a autoclosable construction, to avoid the need to call start and stop. will result in cleaner code.

---

*[src/games/performance/PerformanceConsole.java, line 11 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n43AMzqrSQ2rqgwc8:-K8n43AMzqrSQ2rqgwc9:-1304826034) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/performance/PerformanceConsole.java#L11)):*
when it comes to the Singleton pattern in java I strongly prefer the enum approach, see http://javarevisited.blogspot.hu/2012/07/why-enum-singleton-are-better-in-java.html

---

*[src/games/performance/PerfTimer.java, line 5 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n4VHT6204okNXWUFX:-K8n4VHT6204okNXWUFY:-1064034138) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/performance/PerfTimer.java#L5)):*
again would prefer some autoclosable approach

---

*[src/games/strategy/debug/ErrorConsole.java, line 14 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n4_OOeYsy9C-8G9WG:-K8n4_OOeYsy9C-8G9WH:-530668247) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/strategy/debug/ErrorConsole.java#L14)):*
again make it enum

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/335)*

**Comment by DanVanAtta** [comment]
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/games/performance/Perf.java, line 41 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n3miiXwk5Nkj6WXsK:-K8zxzxt_I9DXTTXJi-9:1670746916) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/performance/Perf.java#L41)):*
I would need to be shown an example I think to really be convinced. I'm concerned the try block indentation will make things difficult, but most importantly it could have some negative consequences for scoping. What if we need to continue using a variable that is created after the try block has started, but is then still used even after the close of the try block. We could elevate the scope of those variables, but then we would need to add null checks, and things are really messy then (not to mention the extra indentation). It's an interesting idea, but I think there are some implications to it that make it not very pragmatic.

---

*[src/games/performance/PerformanceConsole.java, line 11 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n43AMzqrSQ2rqgwc8:-K8zz9cBpaZA-8m_EeGq:37477910) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/performance/PerformanceConsole.java#L11)):*
Preferences aside, my concern is lack of consistency for how singletons are done, but notably here, the class is based on existing singletons. I would opt to update more of them, then have this one be an isolated example of doing singletons in a different way.

Fundamentally, singletons are largely considered an anti-pattern, I'd  rather spend the time removing the singleton implementations altogether TBH.

---

*[src/games/strategy/debug/ErrorConsole.java, line 14 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/335#-K8n4_OOeYsy9C-8G9WG:-K9--fGbDZoBcgT0pEGZ:-783459823) ([raw file](https://github.com/triplea-game/triplea/blob/1c926e041cf239b504839c482b7204ff12ab4a50/src/games/strategy/debug/ErrorConsole.java#L14)):*
Since this is a file move, I want to keep the diff cleaner so it can be reviewed easily. The double checked locks in the article you sent (http://javarevisited.blogspot.hu/2012/07/why-enum-singleton-are-better-in-java.html), is not a benefit - double checked locks atually don't work in Java (has to do with JVM re-ordering of statements and when code crosses read/write barriers).

With that said, I like the benefits pointed out in the article of:
"2) Enum Singletons handled Serialization by themselves"
"3) Creation of Enum instance is thread-safe"

With that, I would like to leave this PR in its current state to facilitate the diff. I'll create a ticket to track updates to the singletons. I could see a small project being dedicated to fixing those up in mass.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/335)*

**Comment by DanVanAtta** [comment]
@gaborbernat , do you still think auto-closable blocks would simplify things? I'm truly not convinced, but not convinced the other way necessarily either.

**Comment by gaborbernat** [comment]
I think making it auto-closable still allows for the current working while offering the option of using the timer as a resource.

**Comment by ron-murhammer** [comment]
@DanVanAtta Tested this out and everything seems to work except 'Enumerate Threads' doesn't show anything in the performance window but instead dumps it to the IDE console?

File: src/games/performance/Perf.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    final Preferences prefs = Preferences.userNodeForPackage(EnablePerformanceLoggingCheckBox.class);
+    prefs.put(LOG_PERFORMANCE_KEY, Boolean.valueOf(enabled).toString());
+  }
+
+  public static boolean isEnabled() {
+    final Preferences prefs = Preferences.userNodeForPackage(EnablePerformanceLoggingCheckBox.class);
+    return prefs.getBoolean(LOG_PERFORMANCE_KEY, false);
+  }
+
+  public static PerfTimer startTimer(Object obj, String title) {
```
Comment (ron-murhammer) [inline src/games/performance/Perf.java]: It doesn't appear that the startTimer and stopTimer methods are used anywhere? These and the PerfTimer are all dead code right now? Do we have an immediate plan to use these?

File: src/games/performance/EnablePerformanceLoggingCheckBox.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import java.awt.event.KeyEvent;
+import javax.swing.JCheckBoxMenuItem;
+
+public class EnablePerformanceLoggingCheckBox extends JCheckBoxMenuItem {
+  private static final long serialVersionUID = -8622339162927764406L;
+
+  public EnablePerformanceLoggingCheckBox() {
+    super("Enable Performance Logging");
+    setSelected(Perf.isEnabled());
+    setMnemonic(KeyEvent.VK_P);
```
Comment (ron-murhammer) [inline src/games/performance/EnablePerformanceLoggingCheckBox.java]: I think 'P' is already used for Politics panel. Did you test this?

File: src/games/performance/PerfTimer.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+public class PerfTimer {
+
+  protected static final PerfTimer DISABLED_TIMER = new PerfTimer("disabled");
+
+  public final String title;
+  private final long startMillis;
+
+  public PerfTimer(String title) {
+    this.title = title;
+    this.startMillis = System.currentTimeMillis();
```
Comment (ron-murhammer) [inline src/games/performance/PerfTimer.java]: Depending on what you are timing you might need to use nano instead of milli.

**Comment by ron-murhammer** [comment]
@DanVanAtta Added a few comments and questions above. My only other thought is around packages. You appear to have created a 'games.performance' which besides util.image is the only package outside of 'games.strategy'. I think maybe we need to think a little bit about package structure and how we'd like it to look to make sure this fits in.

Personally using 'games.strategy' as the prefix for everything doesn't make much sense. Usually organizations use something like 'com.xxx' or 'org.xxx' as the prefix to its unique. In this case, my thought is all packages should start with something like 'triplea' or 'triplea.game' or something that's somewhat unique to TripleA.

All that being said it might be better to keep this more consistent with the existing structure.so 'games.strategy.performance' would probably be best. Also, having PerformanceLoggingCheckBox in a completely different place than all the other menu options seems strange. Unless we are going to use that in multiple places or look to refactor all of the menu options into separate classes its probably better to keep it with the others in TripleaMenu.

File: src/games/performance/Perf.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/performance/Perf.java]: I've been working on another branch that builds on this, uses this as infrastructure. We've had multiple needs for being able to time things easily. Each time I add an ad-hoc class that is similar to StopWatch that has a simpler API and then delete it afterwards. Having a more advanced framework will help.

But... it does look like that branch did get dropped.. My mistake. Regardless, there are some timers I certainly wouldn't mind adding in the short term. With recent work on downloading maps, map parsing, game loading, having this framework will be of use very quickly.

File: src/games/performance/EnablePerformanceLoggingCheckBox.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/performance/EnablePerformanceLoggingCheckBox.java]: Good catch - Changed to L (4336817), and now verified.

**Comment by DanVanAtta** [comment]
I'd vote for our top level packages to be similar to 'triplea.client', 'triplea.lobby', 'triplea.diceserver', 'triplea.client.performance'. As such.. I moved 'games.performance' to 'games.strategy.performance'.

PerformanceLoggingCheckBox does not belong with TripleaMenu, that would be adding to a big mess. Adding to a mess for the sake of being consistently messy is not a compelling case for me.

**Comment by ron-murhammer** [comment]
@DanVanAtta Alright. As long as we look to use the timers in future PRs relatively soon then I'm good with this. Merging now.

**Pull request merged**